### ISSUE TITLE
incorrect diagnostic on `!` async function: expected !, found ()

### ISSUE THREAD

**Issue by njaard** [rust-lang/rust-analyzer#22847]
When I run the following basic lib, which builds cleanly with rustc:

rustc: 1.71.1
rust-analyzer: 1.97.1
editor: Zed

Appears to be a regression from 1.69.

Code to reproduce:

```rust
async fn test() -> ! {
    loop {}
}

pub async fn test1() -> ! {
    test().await;
}
^ expected !, found () (rust-analyzer E0308)
```

If I add an `unreachable!` before the diagnostic, it curiously says that it's unreachable.

**Comment by njaard** [comment]
@ChayimFriedman2 My exact test case above continues to produce that incorrect warning as of rust-analyzer 1.98.1 (48a229c 2026-09-01).

**Comment by ChayimFriedman2** [comment]
That's because the fix is only included in 1.99.0.
